File: lib/hyprdesk/theme.py
"""Central config + theme resolution for all python desktop widgets."""
import re
from pathlib import Path

HOME = Path.home()
WIDGETS_CONF = HOME / ".config/conky/widgets.conf"
WALLUST_COLORS = HOME / ".config/conky/colors.lua"
THEMES_DIR = HOME / ".config/conky/themes"

FALLBACK = {"bg": "#101017", "fg": "#FEFAD7", "accent": "#F2E3EF",
            "accent2": "#FCF18E", "muted": "#383940"}


def conf():
    """widgets.conf as a dict — the single settings file for the fleet."""
    d = {}
    try:
        for line in WIDGETS_CONF.read_text().splitlines():
            m = re.match(r"^(\w+)\s*=\s*(\S+)", line)
            if m:
                d[m.group(1)] = m.group(2)
    except OSError:
        pass
    return d


def conf_get(key, default=""):
    return conf().get(key, default)


def colors():
    """Active palette: theme=wallust follows the wallpaper, any other
    name loads ~/.config/conky/themes/<name>.lua."""
    src = WALLUST_COLORS
    theme = conf_get("theme", "wallust")
    if theme != "wallust":
        themed = THEMES_DIR / f"{theme}.lua"
        if themed.is_file():
            src = themed
    pal = dict(FALLBACK)
    try:
        for k, v in re.findall(r'(\w+)\s*=\s*"(#[0-9a-fA-F]{6})"',
                               src.read_text()):
            pal[k] = v
    except OSError:
        pass
    # derived ink hierarchy — keep in sync with card.lua M.colors()
    pal.setdefault("sub", _mix(pal["fg"], pal["bg"], 0.62))
    # STATUS INK IS PINNED, NOT WALLPAPER-DERIVED.
    # wallust maps wallpaper colours into ansi slots whose NAMES are a lie
    # (@red has been a dark teal at 1.00:1 against the background). A state
    # the user must be able to read cannot be left to that lottery, so
    # good/bad/warn are constants everywhere on the machine — see
    # ~/.config/wallust/templates/*.css, which emit the same three values
    # into the GTK consumers.
    pal.setdefault("good", "#8EC07C")   # 8.07:1 on the current ground
    pal.setdefault("bad", "#E06C75")    # 5.31:1
    pal.setdefault("warn", "#E0B25C")   # 9.06:1
    return pal


def _mix(a, b, t):
    ar, ag, ab = (int(a[i:i + 2], 16) for i in (1, 3, 5))
    br, bg, bb = (int(b[i:i + 2], 16) for i in (1, 3, 5))
    return "#%02X%02X%02X" % (round(ar * t + br * (1 - t)),
                              round(ag * t + bg * (1 - t)),
                              round(ab * t + bb * (1 - t)))


def rgb(hexstr):
    """'#RRGGBB' -> (r, g, b) floats 0..1 for cairo."""
    return tuple(int(hexstr[i:i + 2], 16) / 255 for i in (1, 3, 5))


def css_rgba(hexstr, alpha):
    """'#RRGGBB' -> 'rgba(r,g,b,a)' for GTK CSS."""
    r, g, b = (int(hexstr[i:i + 2], 16) for i in (1, 3, 5))
    return f"rgba({r},{g},{b},{alpha})"

File: rust/hyprdesk/src/theme.rs
//! theme.py's colors(), full palette — including the PINNED status inks
//! (status colours never ride the wallpaper) and both contrast fixes.

use std::fs;

#[derive(Clone, Copy, Debug, PartialEq)]
pub struct Rgb(pub u8, pub u8, pub u8);

impl Rgb {
    pub fn parse(s: &str) -> Option<Rgb> {
        let h = s.strip_prefix('#')?;
        if h.len() != 6 {
            return None;
        }
        Some(Rgb(
            u8::from_str_radix(&h[0..2], 16).ok()?,
            u8::from_str_radix(&h[2..4], 16).ok()?,
            u8::from_str_radix(&h[4..6], 16).ok()?,
        ))
    }
    fn lum(self) -> f64 {
        let f = |c: u8| {
            let x = c as f64 / 255.0;
            if x <= 0.03928 {
                x / 12.92
            } else {
                ((x + 0.055) / 1.055).powf(2.4)
            }
        };
        0.2126 * f(self.0) + 0.7152 * f(self.1) + 0.0722 * f(self.2)
    }
    pub fn ratio(a: Rgb, b: Rgb) -> f64 {
        let (la, lb) = (a.lum(), b.lum());
        let (hi, lo) = (la.max(lb), la.min(lb));
        (hi + 0.05) / (lo + 0.05)
    }
    pub fn mix(a: Rgb, b: Rgb, t: f64) -> Rgb {
        let m = |x: u8, y: u8| (x as f64 + (y as f64 - x as f64) * t).round() as u8;
        Rgb(m(a.0, b.0), m(a.1, b.1), m(a.2, b.2))
    }
}

#[derive(Clone, Copy, Debug)]
pub struct Palette {
    pub bg: Rgb,
    pub fg: Rgb,
    pub accent: Rgb,
    pub accent2: Rgb,
    pub sub: Rgb,
    pub muted: Rgb,
    pub good: Rgb,
    pub bad: Rgb,
    pub warn: Rgb,
}

pub fn colors() -> Palette {
    let mut bg = Rgb::parse("#101017").unwrap();
    let mut fg = Rgb::parse("#FEFAD7").unwrap();
    let mut accent = Rgb::parse("#F2E3EF").unwrap();
    let mut accent2 = Rgb::parse("#FCF18E").unwrap();
    let mut muted = Rgb::parse("#3D3B3A").unwrap();
    // sub comes FROM the file when wallust provides it (it does on this
    // box); the mix() below is only the fallback. Deriving unconditionally
    // read #6D6C60 where python read #A5A391 — caught by the parity test.
    let mut sub: Option<Rgb> = None;
    // status inks are pinned but THEME-OVERRIDABLE (ink table): a theme
    // file may redefine them, the wallpaper never does
    let mut good = Rgb::parse("#8EC07C").unwrap();
    let mut bad = Rgb::parse("#E06C75").unwrap();
    let mut warn = Rgb::parse("#E0B25C").unwrap();

    let theme = crate::conf_get("theme", "wallust");
    let mut src = crate::home().join(".config/conky/colors.lua");
    if theme != "wallust" {
        let themed = crate::home().join(format!(".config/conky/themes/{theme}.lua"));
        if themed.is_file() {
            src = themed;
        }
    }
    if let Ok(text) = fs::read_to_string(&src) {
        for line in text.lines() {
            let Some((name, rest)) = line.split_once('=') else {
                continue;
            };
            let name = name.trim();
            let Some(hexpos) = rest.find('#') else { continue };
            let hex: String = rest[hexpos..].chars().take(7).collect();
            if let Some(c) = Rgb::parse(&hex) {
                match name {
                    "bg" => bg = c,
                    "fg" => fg = c,
                    "accent" => accent = c,
                    "accent2" => accent2 = c,
                    "muted" => muted = c,
                    "sub" => sub = Some(c),
                    "good" => good = c,
                    "bad" => bad = c,
                    "warn" => warn = c,
                    _ => {}
                }
            }
        }
    }
    // A TIER THAT IS NOT DISTINCT IS NOT A TIER (theme.py): accent2 rides
    // the wallpaper and can land indistinguishable from fg.
    if Rgb::ratio(accent2, fg) < 1.6 {
        accent2 = accent;
    }
    // muted is a SURFACE tone; a wallpaper can hand one invisible on bg
    if Rgb::ratio(muted, bg) < 1.35 {
        muted = Rgb::mix(bg, fg, 0.22);
    }
    Palette {
        bg,
        fg,
        accent,
        accent2,
        // python's _mix(fg,bg,t) weights its FIRST argument: bg + (fg-bg)*t.
        // Rgb::mix weights the second, so the arguments swap here — the
        // parity test read #6D6C60 against python's #A5A391 until they did.
        sub: sub.unwrap_or_else(|| Rgb::mix(bg, fg, 0.62)),
        muted,
        good,
        bad,
        warn,
    }
}
